doc/man/66-boot.8.scd
66-boot(8)

# NAME
66-boot - Run as pid 1 and prepare the system for scandir

# SYNOPSIS

66-boot [ *-h* ] [ *-s* _skel_ ] [ *-m* ] [ *-l* _log_user_ ] [ *-e* _environment_ ] [ *-d* _dev_ ] [ *-b* _banner_ ]

# DESCRIPTION

66-boot runs as pid 1, as a _stage1_ init; it performs the necessary early system preparation spawns a process that will run the _rc.init_ script, then execs into *66-scandir*(1).

# OPTIONS 

*-h*
	Prints this help.

*-s* _skel_
	Path of the skeleton files. By default this will be *%%skel%%*.
	The default can also be changed at compile time by passing the
	*--with-skeleton=*_DIR_ option to *./configure*. This directory *must*
	contains the necessaries skeleton files to properly boot the machine, without it the system *will not boot*. _skel_ must be an absolute path.

*-m*
	Umount the basename of the _LIVE_ directory set into the _init.conf_
	skeleton file and mount a tmpfs on it. This is the default.
	Umount the basename of the _LIVE_ directory set into the _init.conf_
	skeleton file if it already mounted and mount a tmpfs on it. By default, 
	the _LIVE_ basename is mounted if it not already a valid mountpoint. 
	Otherwise without the *-m* option, it does nothing.

*-l* _log_user_
	The catch-all logger will run as the _log_user_ user. Default is *%%s6log_user%%*.
	The default can also be changed at compile-time by passing the
	\--with-s6-log-user=_user_ option to *./configure*.

*-e* _environment_
	_stage 1_ init empties its environment before spawning _rc.init_ skeleton
	file and executing into *66-scandir*(1), in order to prevent those "kernel"
	environment variables from leaking into the whole process tree. The *PATH*
	variable will be the only one variable set into the environment. If you
	want to define environment variable use this options. _environment_ must be
	an absolute path and reacts at the same as *66-scandir*(1) *-e* option.

*-d* _dev_
	Mount a devtmpfs on */dev*. By default, no such mount is performed - it is
	assumed that a devtmpfs is automounted on */dev* at boot time by the kernel
	or an initramfs.

*-b* _banner_
	Prints _banner_ to */dev/console* at the start of the _stage 1_ init
	process. By default "[Starts stage1 process ...]".

# EARLY PREPARATION

When booting a system, *66-boot* performs the following operations:

- It prints a banner to */dev/console*.
- It parse the _init.conf_ skeleton file.
- It chdirs to */*.
- It sets the umask to _initial_umask_.
- It becomes a session leader.
- It mounts a devtmpfs on */dev*, if requested.
- It uses */dev/null* as its stdin (instead of */dev/console*).
	- *dev/console* is still used, for now, as stdout and stderr.
- It checks if the _LIVE_ basename is a valid mountpoint and if so mounts it.
  If requested, it unmounts if the _LIVE_ basename is a valid mountpoint
  and performs a mount.
- It create _LIVE_ directory invocating
  "66-scandir -v _VERBOSITY_ -l _LIVE_ -b -c -s _skel_" more "-L user_log" if
  requested.
- It initiate the early services of _TREE_ invocating
  "66-init -v VERBOSITY -l LIVE -t TREE classic".
- It reads the initial environment from _environment_ if requested.
- It performs "the fifo trick", i.e. it redirects its stdout to the catch-all
  logger's fifo, without blocking, before the catch-all logger is even up
  (because it's a service that will be spawned a bit later, when
  *66-scandir*(1) is executed).
- It forks a child, also called _stage2_:
	- The child blocks until the catch-all logger runs.
	- The child starts service of tree _TREE_.
	- The child becomes a session leader.

- It also makes the catch-all logger's fifo its stderr.
- It execs into "66-scandir -v VERBOSITY -l LIVE -u" with *LIVE/scandir/0*
  (default */run/66/scandir/0*) as its scandir.

- *66-scandir*(1) pass the relay to *s6-svscan* which spawns the early
  services that are defined in _TREE_ where one of those services is
  _scandir-log_, which is the catch-all logger. When this service is up,
  *66-boot*'s child unblocks.
- The child execs into _rc.init_.

If one of the above processus fail, *66-boot* try to launch a Single-user login
a.k.a. _sulogin_ which let you the ability to fix the error. This case should
not arrive.

# EXIT CODES

*66-boot* never exits. It spawns the _rc.init_ script and execs
*66-scandir*(1), which runs forever until the machine stops or reboots.

# SKELETON FILES

Skeleton files are mandatories and must exist on your system to be able to boot
and shutdown the machine properly. By default those files are installed at
*%%skel%%*. Use the *--with-skeleton=*_DIR_ option at compile time to change
it.

- *init*++
*66-boot* binary is not meant to be called directly, or be linked to the
  binary directory directly, because it takes command-line options. Thereby
  the _init_ file is used to pass an options to *66-boot*. By default
  *66-boot* is launched with *-m* options. This file *should be copied* by
  the administrator into the binary directory of your system.

- *init.conf*++
this file contains a set of _key_*=*_value_ pairs. *All* keys are mandatory
  and thereby the name of the key *cannot* be changed. This is the file used
  configure the boot process. By default:

	- *VERBOSITY=*_0_++
increase/decrease the verbosity of _stage1_ process.

	- *LIVE=*_/run/66_++
create the scandir at _LIVE_. The value by default will depend of
	  the *--livedir=*_live_ set a compile time. An absolute is expected
	  as value.

	- *PATH=*_/usr/bin:/usr/sbin:/bin:/sbin:/usr/local/bin_++
the initial value for the *PATH* environment variable, that will be
	  transmitted to all the starting process unless it's overridden by a
	  *PATH* declaration via the *-e* option. It is absolutely necessary
	  for *execline*, *s6*, *s6-rc* and all *66 tools* binaries to be
	  accessible via *PATH*, else the machine will not boot.

	- *TREE=*_init_++
name of the _tree_ to start. This _tree_ should contain a set of
	  service to bring up the machine to an operating system. 'classic'
	  services will become earlier service at *66-init* invocation.
	  _stage2_ will start all other kind of service defined into it. It's the
	  responsability of the administrator to correctly set this tree.

	- *RCINIT=*%%skel%%/rc.init_++
This file is launched at the end of the _stage1_ and run as _stage2_.
	  It call *66-init*(1) to iniatiate all service of _TREE_ except
	  'classic' which are already initiated on the _stage1_ than invoke
	  *66-dbctl*(1) to bring up the services. An absolute path is expected as
	  value pointing to the name of the file to run.

	- *RCSHUTDOWN=*%%skel%%/rc.shutdown_++
This is launched when a shutdown is requested also called _stage3_. It invoke
	  *66-all*(1) to bring down all services of _TREE_. An absolute path is
	  expected as value pointing to the name of the file to run.

	- *RCSHUTDOWNFINAL=*%%skel%%/rc.shutdown_++
This file will be run at the very end of the shutdown procedure, after 
      all processes have been killed and all filesystems have been unmounted,
      just before the system is rebooted or the power turned off.
	
	- *UMASK=*_0022_++
Set the value of the initial file umask for all the starting processes, in
	  octal.

	- *RESCAN=*_0_++
Ask at *s6-svscan* to perform a scan every _RESCAN_ milliseconds. It should
	  be 0 at _stage1_ but its here just in case. It is strongly discouraged to
	  set _RESCAN_ to a positive value under 500.

	- *ISHELL=*%%skel%%/ishell_++
Run _ISHELL_ in case of _stage2_ crash. This file try to run a *sulogin*. An
	  absolute path is expected as value pointing to the name of the file to run.

- *rc.init*++
This file is called by the child of *66-boot* to process the _stage2_. It
  invoke two command:

	- *66-init -v${VERBOSITY} -l ${LIVE} -t ${TREE} database* will initiate
	  *bundle* and *atomic* service of the _TREE_. If the process fail, it try
	  to run _ISHELL_ as command to rescue.

	- *66-dbctl -v${VERBOSITY} -l ${LIVE} -t ${TREE} -u* will bring up all
	  *bundle* and *atomic* service of the _TREE_. If the process fail, it try
	  to run _ISHELL_ as command to rescue.

- *rc.shutdown*++
This file is called at shutdown request when the administrator runs the
  *shutdown*, *halt*, *poweroff* or *reboot* command.++
It invoke one command:

	- *66-all -v${VERBOSITY} -l ${LIVE} -t ${TREE} -f down* will bring down all
	  _services_ for all _tree_ marked enabled.

- *ishell*++
this file is called by *rc.init* in case of crash of the _stage2_ and try to
  run a *sulogin*.

- *halt*++
Safe wrapper around *66-hpr -h* command and accept any command options of it.
  This is the file called to halt the system and *should be copied or
  symlinked* by the administrator into the binary directory of your system.


- *poweroff*++
Safe wrapper around *66-hpr -p* command and accept any command options of it.
  This is the file called to poweroff the system and *should be copied or
  symlinked* by the administrator into the binary directory of your system.

- *reboot*++
Safe wrapper around *66-hpr -r* command and accept any command options of it.
  This is the file called to reboot the system and *should be copied or
  symlinked* by the administrator into the binary directory of your system.

- *shutdown*++
Safe wrapper around *66-shutdown*(1) command and accept any command options of
  it. This is the file called to shutdown the system and *should be copied or
  symlinked* by the administrator into the binary directory of your system.


NEWS
Changelog for 66

In 0.2.5.2
----------

    - Fix build: Remove bytes.h oblibs header file
    
In 0.2.5.1
----------
    
    - Bugs fix
    - 66-tree -S options: if after_tree and tree have the same name,
    tree is considered as the very first tree to start.
    - 66-unmountall: do not umount SS_LIVE
    - Add SIGPWR control file at creation of .s6-svscan directory

In 0.2.5.0
----------
    
    - Bugs Fix: Bad memory allocation
    - 66-tree: add -S option which allow to order the start tree process
    even after it creation.
    
In 0.2.4.1
----------
    
    - Hot fix: fix html documentation

In 0.2.4.0
----------
    
    - Bugs fix: Bad memory allocation.
    
    - Enable again RB_DISABLE_CAD but don't crash if its not available
    and warn user
    
    - Add @optsdepends and @extdepends field at [main] section:
        - @optsdepends can be considere as: "enable one on this service 
        or none"
        - @extdepends can be considere as: "enable the service if it is 
        not already declared on a tree"
    
    - 66-in{tree,service}: 
        - in case of empty value the tools return None
        - add -n option: this avoids to display the name of the field.
        - add field at 66-intree:
            - start: displays the list of tree(s) started before
            - allowed: displays a list of allowed user to use the tree
            - symlinks: displays the target of tree's symlinks
        - add field at 66-inservice:
            - optsdepends: displays the service's optional dependencies
            - extdepends: displays the service's external dependencies
    
    - 66-shutdown skeleton: be safier and check if options are past
    
    - New tool:
        - 66-update: The 66-update program makes a complete transition 
        of trees and the live directory using a old 66 format 
        (the one being replaced) with the new 66 format.

In 0.2.3.2
----------
    
    - Fix bad annoucement at 66-svctl
    - Warn in case of bad key at parenthesis parsing process
    
In 0.2.3.1
----------
    
    - Bugs fix: bad allocation memory
    - Makefile: remove creation of empty directories. 66-tree will check
    all directories and create it if it missing.
    
In 0.2.3.0
----------
    
    - Adapt to oblibs 0.0.4.0
    - Remove RB_DISABLE_CAD at 66-boot by default and allow to enable
    it with -c option. Starting on virtual system like lxc crash with
    this option. Well be safe on every case.
    - Use of new function log_? familly from oblibs:
        - standardization of the exit code
        - standardization of the verbosity output
        - allow verbosity to 4 to display debug message
    - Doc typo fix        
    
In 0.2.2.2
----------
    
    - Adapt to skalibs 2.9.1.0
    - Remove insta_? deprecated function
    
In 0.2.2.1
----------
    
    - Typo fix at html documentation
    - Fix 66-intree without options
    
In 0.2.2.0
----------

    - Bugs fix: Always check the existing of the 66 heart directories
    - 66-info is now deprecated and splitted into two differents API.
    66-intree give informations about tree
    66-inservice give informations about services

In 0.2.1.2
----------

    - Hot fix: fix 66-info graph display.
    
In 0.2.1.1
----------

    - Hot fix: fix the build of the service dependencies graph.

In 0.2.1.0
----------

    - Bugs fix.
   
    - A synchronization is now made on reboot even with force option.
   
    - The field @name has no effects except for instantiated service but 
    can be omitted. The name of the frontend file is took as name. 
    In case of instantiated service, the field @name must contain the 
    complete name of the frontend service file -- Refer to the frontend 
    documentation page.
    
    - New skeleton file: rc.shutdown.final. This skeleton file will be run
    at the very end of the shutdown procedure, after all processes have 
    been killed and all filesystems have been unmounted, just before the 
    system is rebooted or the power turned off.
    This script normally remains empty.
   
    - Extra tools has removed and provided as an independent programs at
    https://framagit.org/Obarun/66-tools.git except for 66-echo and 
    66-unmountall program which are a dependent part of 66.
   
    - The account to run the s6-log program at the associate service logger
    can be set at compilation time by the --with-s6-log-user=USER options.
    The default is root. Obviously, the @runas field at the [logger] section
    overwritte it.
    This options set the account to run the uncaught-logs too. Also,
    this option can be overwritten with the '-l' option at 66-boot invocation
    and the '-L' option at 66-scandir invocation. 
    
In 0.2.0.4
----------
	
	- Bugs fix on memory allocation
	- Respect policies decision of user: 
		Remove -m option from 66-boot on init file
	- Add -d feature to 66-scandir 

In 0.2.0.3
----------

	- 66-parser: Fix write of configuration file, add -c|C features 
	
In 0.2.0.2
----------

	- Fix the written of user configuration file 
	- Fix oneshot and bundle status check
	- Add -l features to 66-tree, rewrite tree_unsupervise function
	
In 0.2.0.0
----------

	- New tools:
		- 66-boot.
		- 66-scanctl.
		- 66-shutdown.
		- 66-shutdownd.
		- 66-hpr.
		- 66-env.
		- 66-parser.
		- 66-which.
		- 66-echo.
		- 66-unmountall.
		- execl-subuidgid.
	- New @key field and change on frontend file:
		- [Logger] section accept a @depends field, @timestamp accept none
		as value, readiness notification is a default.
		- @hiercopy in [main] allow to copy any file or directory coming
		from the directory of the service.
	- 66-envfile is now deprecated, use execl-envfile in replacement.
	- 66-enable: add -F, -c, -C features.
	- 66-svctl is now asynchrone to launch services.
	- Man pages are available.
	- Environment variables limitation: maximum 100 files by directory,
	each file cannot contain more than 4096 bytes or 50 variables.

src/66/66-boot.c
/* 
 * 66-boot.c
 * 
 * Copyright (c) 2018-2020 Eric Vidal <eric@obarun.org>
 * 
 * All rights reserved.
 * 
 * This file is part of Obarun. It is subject to the license terms in
 * the LICENSE file found in the top-level directory of this
 * distribution.
 * This file may not be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file./
 */
 
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/stat.h>
#include <sys/mount.h>
#include <sys/reboot.h>

#include <oblibs/log.h>
#include <oblibs/files.h>
#include <oblibs/string.h>
#include <oblibs/obgetopt.h>
#include <oblibs/environ.h>
#include <oblibs/sastr.h>

#include <skalibs/buffer.h>
#include <skalibs/djbunix.h>
#include <skalibs/stralloc.h>
#include <skalibs/types.h>

#include <66/config.h>
#include <66/constants.h>

static mode_t mask = SS_BOOT_UMASK ;
static unsigned int rescan = SS_BOOT_RESCAN ;
static char const *skel = SS_SKEL_DIR ;
static char *live = SS_LIVE ;
static char const *path = SS_BOOT_PATH ;
static char const *tree = SS_BOOT_TREE ;
static char const *rcinit = SS_SKEL_DIR SS_BOOT_RCINIT ;
static char const *banner = "\n[Starts stage1 process...]" ;
static char const *slashdev = 0 ;
static char const *envdir = 0 ;
static char const *fifo = 0 ;
static char const *log_user = SS_LOGGER_RUNNER ;
static char const *cver = 0 ;
static char tpath[MAXENV+1] ;
static char trcinit[MAXENV+1] ;
static char tlive[MAXENV+1] ;
static char ttree[MAXENV+1] ;
static char confile[MAXENV+1] ;
static char const *const *genv = 0 ;
static int fdin ;

#define USAGE "66-boot [ -h ] [ -m ] [ -s skel ] [ -l log_user ] [ -e environment ] [ -d dev ] [ -b banner ]"

static void sulogin(char const *msg,char const *arg)
{
	static char const *const newarg[2] = { SS_EXTBINPREFIX "sulogin" , 0 } ;
	pid_t pid ;
	int wstat ;
	fd_close(0) ;
	if (dup2(fdin,0) == -1) log_dieu(LOG_EXIT_SYS,"duplicate stdin -- you are on your own") ;
	fd_close(fdin) ;	
	if (*msg) log_warnu(msg,arg) ;
	pid = child_spawn0(newarg[0],newarg,genv) ;
	if (waitpid_nointr(pid,&wstat, 0) < 0)
			log_dieusys(LOG_EXIT_SYS,"wait for sulogin -- you are on your own") ;
	fdin=dup(0) ;
	if (fdin == -1) log_dieu(LOG_EXIT_SYS,"duplicate stdin -- you are on your own") ;
	fd_close(0) ;
	if (open("/dev/null",O_WRONLY)) log_dieu(LOG_EXIT_SYS,"open /dev/null -- you are on your own") ;
}

static inline void info_help (void)
{
  static char const *help =
"66-boot <options>\n"
"\n"
"options :\n"
"	-h: print this help\n" 
"	-m: mount parent live directory\n"
"	-l: run catch-all logger as log_user user\n"
"	-s: skeleton directory\n"
"	-e: environment directory or file\n"
"	-d: dev directory\n"
"	-b: banner to display\n"
;

 if (buffer_putsflush(buffer_1, help) < 0) sulogin("","") ;
}

static void parse_conf(void)
{
	static char const *valid[] = 
	{ "VERBOSITY", "PATH", "LIVE", "TREE", "RCINIT", "UMASK", "RESCAN", 0 } ;
	int r ;
	unsigned int j = 0 ;
	stralloc src = STRALLOC_ZERO ;
	stralloc val = STRALLOC_ZERO ;
	if (skel[0] != '/') sulogin("skeleton directory must be an aboslute path: ",skel) ;
	size_t skelen = strlen(skel) ;
	memcpy(confile,skel,skelen) ;
	confile[skelen] = '/' ;
	memcpy(confile + skelen + 1, SS_BOOT_CONF, SS_BOOT_CONF_LEN) ;
	confile[skelen + 1 + SS_BOOT_CONF_LEN] = 0 ;
	size_t filesize=file_get_size(confile) ;
	r = openreadfileclose(confile,&src,filesize) ;
	if(!r) sulogin("open configuration file: ",confile) ; 
	if (!stralloc_0(&src)) sulogin("append stralloc of file: ",confile) ;
	
	for (char const *const *p = valid;*p;p++,j++)
	{
		if (!stralloc_copy(&val,&src)) sulogin("copy stralloc of file: ",confile) ;
		if (!environ_get_val_of_key(&val,*p)) continue ;
		/** value may be empty, in this case we use the default one */
		if (!sastr_clean_element(&val)) 
		{	
			log_warnu("get value of: ",*p," -- keeps the default") ;
			continue ;
		}
		if (!sastr_rebuild_in_oneline(&val)) sulogin("rebuild line of value: ",val.s) ;
		if (!stralloc_0(&val)) sulogin("append stralloc of value: ",val.s) ;
		
		switch (j)
		{
			case 0: if (!uint0_scan(val.s, &VERBOSITY)) sulogin("parse VERBOSITY value: ",val.s) ; 
					break ;
			case 1: memcpy(tpath,val.s,val.len) ;
					tpath[val.len] = 0 ;
					path = tpath ;
					break ;
			case 2: memcpy(tlive,val.s,val.len) ;
					tlive[val.len] = 0 ;
					live = tlive ;
					if (live[0] != '/') sulogin ("LIVE must be an absolute path: ",val.s) ; 
					break ;
			case 3: memcpy(ttree,val.s,val.len) ;
					ttree[val.len] = 0 ;
					tree = ttree ;
					break ;
			case 4: memcpy(trcinit,val.s,val.len) ;
					trcinit[val.len] = 0 ;
					rcinit = trcinit ;
					if (rcinit[0] != '/') sulogin ("RCINIT must be an absolute path: ",val.s) ; 
					break ;
			case 5: if (!uint0_oscan(val.s, &mask)) sulogin("invalid MASK value: ",val.s) ; break ;
			case 6: if (!uint0_scan(val.s, &rescan)) sulogin("invalid RESCAN value: ",val.s) ; break ;
			default: break ;
		}
		
	}
	stralloc_free(&val) ;
	stralloc_free(&src) ;
}

static int is_mnt(char const *str)
{
	struct stat st;
	size_t slen = strlen(str) ;
	int is_not_mnt = 0 ;
	if (lstat(str,&st) < 0) sulogin("lstat: ",str) ;
	if (S_ISDIR(st.st_mode))
	{
		dev_t st_dev = st.st_dev ; ino_t st_ino = st.st_ino ;
		char p[slen+4] ;
		memcpy(p,str,slen) ;
		memcpy(p + slen,"/..",3) ;
		p[slen+3] = 0 ;
		if (!stat(p,&st))
			is_not_mnt = (st_dev == st.st_dev) && (st_ino != st.st_ino) ;
	}else return 0 ;
	return is_not_mnt ? 0 : 1 ;
}

static void split_tmpfs(char *dst,char *str)
{
	size_t len = get_len_until(str+1,'/') ;
	len++ ;
	memcpy(dst,str,len) ;
	dst[len] = 0 ;
}

static inline void run_stage2 (char const *const *envp, size_t envlen, char const *modifs, size_t modiflen)
{
	char const *newargv[3] = { rcinit, confile, 0 } ;
	setsid() ;
	fd_close(1) ;
	if (open(fifo, O_WRONLY) != 1)  /* blocks until catch-all logger is up */
		sulogin("open for writing fifo: ",fifo) ;
	if (fd_copy(2, 1) == -1)
		sulogin("copy stderr to stdout","") ;
	fd_close(fdin) ;
	xpathexec_r(newargv, envp, envlen, modifs, modiflen) ;
}
static inline void run_cmdline(char const *const *newargv, char const *const *envp, char const *msg,char const *arg)
{
	pid_t pid ;
	int wstat ;
	pid = child_spawn0(newargv[0],newargv,envp) ;
	if (waitpid_nointr(pid,&wstat, 0) < 0)
		sulogin("wait for: ",newargv[0]) ;
	if (wstat) sulogin(msg,arg) ;
}

static inline void make_cmdline(char const *prog,char const **add,int len,char const *msg,char const *arg,char const *const *envp)
{
	
	int m = 6 + len, i = 0, n = 0 ;
	char const *newargv[m] ;
	newargv[n++] = prog ;
	newargv[n++] = "-v" ; 
	newargv[n++] = cver ;
	newargv[n++] = "-l" ;
	newargv[n++] = live ; 
	for (;i<len;i++)
		newargv[n++] = add[i] ;
	newargv[n] = 0 ;
	run_cmdline(newargv,envp,msg,arg) ;
}

int main(int argc, char const *const *argv,char const *const *envp)
{
	unsigned int r , tmpfs = 0 ;
	size_t bannerlen, livelen ;
	pid_t pid ;
	int opened = 0 ;
	char verbo[UINT_FMT] ;
	cver = verbo ;
	stralloc envmodifs = STRALLOC_ZERO ;
	genv = envp ;
	PROG = "66-boot" ;
	{
		subgetopt_t l = SUBGETOPT_ZERO ;

		for (;;)
		{
			int opt = getopt_args(argc,argv, ">hms:e:d:b:l:", &l) ;
			if (opt == -1) break ;
			if (opt == -2) sulogin("options must be set first","") ;
			switch (opt)
			{
				case 'h' : info_help(); return 0 ;
				case 'm' : tmpfs = 1 ; break ;
				case 's' : skel = l.arg ; break ;
				case 'e' : envdir = l.arg ; break ;
				case 'd' : slashdev = l.arg ; break ;
				case 'b' : banner = l.arg ; break ;
				case 'l' : log_user = l.arg ; break ;
				default :  log_usage(USAGE) ; 
			}
		}
		argc -= l.ind ; argv += l.ind ;
	}
	if (geteuid())
	{
		errno = EPERM ;
		log_diesys(LOG_EXIT_USER, "nice try, peon") ;
	}
	fdin=dup(0) ;
	parse_conf() ;
	verbo[uint_fmt(verbo, VERBOSITY)] = 0 ;
	bannerlen = strlen(banner) ;
	livelen = strlen(live) ;
	char tfifo[livelen + 1 + SS_BOOT_LOGFIFO_LEN + 1] ;
	memcpy(tfifo,live,livelen) ;
	tfifo[livelen] = '/' ;
	memcpy(tfifo + livelen + 1,SS_BOOT_LOGFIFO,SS_BOOT_LOGFIFO_LEN) ;
	tfifo[livelen + 1 + SS_BOOT_LOGFIFO_LEN] = 0 ;
	fifo = tfifo ;
	
	allwrite(1, banner, bannerlen) ;
	allwrite(1, "\n", 2) ;
	if (chdir("/") == -1) sulogin("chdir to ","/") ;
	umask(mask) ;
	setpgid(0, 0) ;
	fd_close(0) ;
	
	if (slashdev)
	{
		log_info("Mount: ",slashdev) ;
		fd_close(1) ;
		fd_close(2) ;
		if (mount("dev", slashdev, "devtmpfs", MS_NOSUID | MS_NOEXEC, "") == -1)
			{ opened++ ; sulogin ("mount: ", slashdev) ; }
		
		if (open("/dev/console", O_WRONLY) ||
		fd_copy(1, 0) == -1 ||
		fd_move(2, 0) == -1) return 111 ;
	}
	if (!opened)
		if (open("/dev/null", O_RDONLY)) sulogin("open: ", "/dev/null") ;
	
	char fs[livelen + 1] ;
	split_tmpfs(fs,live) ;
	r = is_mnt(fs) ;
	
	if (!r || tmpfs)
	{
		if (r && tmpfs)
		{
			log_info("Umount: ",fs) ;
			if (umount(fs) == -1) sulogin ("umount: ",fs ) ;
		}
		log_info("Mount: ",fs) ;
		if (mount("tmpfs", fs, "tmpfs", MS_NODEV | MS_NOSUID, "mode=0755") == -1) 
			sulogin("mount: ",fs) ;
	}
	/** respect the path before run 66-xxx API*/
	if (setenv("PATH", path, 1) == -1) sulogin("set initial PATH: ",path) ;
	/** create scandir */
	{
		char const *t[] = { "-b", "-c", "-s", skel, "-L", log_user } ;
		log_info("Create live scandir at: ",live) ;
		make_cmdline(SS_EXTBINPREFIX "66-scandir",t,6,"create live scandir at: ",live,envp) ;
	}
	/** initiate earlier service */
	{
		char const *t[] = { "-t",tree,"classic" } ;
		log_info("Initiate earlier service of tree: ",tree) ;
		make_cmdline(SS_EXTBINPREFIX "66-init",t,3,"initiate earlier service of tree: ",tree,envp) ;	
	}
	
	if (envdir) {
		int e = environ_get_envfile(&envmodifs,envdir) ;
		if (e <= 0){ environ_get_envfile_error(e,envdir) ; sulogin("","") ; }
	}
	
	{
		log_info("Starts boot logger at: ",live,"/log/0") ;
		int fdr = open_read(fifo) ;
		if (fdr == -1) sulogin("open fifo: ",fifo) ;
		fd_close(1) ;
		if (open(fifo, O_WRONLY) != 1) sulogin("open fifo: ",fifo) ;
		fd_close(fdr) ;
	}
	
	/** fork and starts scandir */
	{
		static char const *newargv[7] ; 
		newargv[0] = SS_EXTBINPREFIX "66-scandir" ;
		newargv[1] = "-v" ;
		newargv[2] = verbo ;
		newargv[3] = "-l" ;
		newargv[4] = live ;
		newargv[5] = "-u" ;
		newargv[6] = 0 ; 
		char const *newenvp[2] = { 0, 0 } ;
		size_t pathlen = strlen(path) ;
		char pathvar[6 + pathlen] ;
		memcpy(pathvar, "PATH=", 5) ;
		memcpy(pathvar + 5, path, pathlen + 1) ;
		newenvp[0] = pathvar ;
		pid = fork() ;
		if (pid == -1) sulogin("fork: ",rcinit) ;
		if (!pid) run_stage2(newenvp, 2, envmodifs.s,envmodifs.len) ;
		if (reboot(RB_DISABLE_CAD) == -1) log_warnusys("trap ctrl-alt-del") ;
		if (fd_copy(2, 1) == -1) sulogin("copy stderr to stdout","") ;
		fd_close(fdin) ;
		xpathexec_r(newargv, newenvp, 2, envmodifs.s, envmodifs.len) ;
	}
}
